src/mlte/value/base.py

```diff
@@ -0,0 +1,111 @@
+"""
+mlte/value/base.py
+
+The base class for MLTE value extensions.
+
+NOTE: The `Value` implementation in this module (`base`) should not be
+confused with the `Value` implementation in the `artifact` module.
+The `Value` implementation in `artifact` is a proper MLTE artifact, it
+should be used for all "internal" types part of the MLTE value system.
+The `Value` implementation in this module is meant to be extended by users
+of MLTE to enrich the value system with their own value types; it provides
+the link between the statically-typed MLTE value system and dynamic extensions.
+"""
+
+from __future__ import annotations
+
+import abc
+import typing
+from typing import Any
+
+from mlte.artifact.artifact import Artifact
+from mlte.artifact.model import ArtifactHeaderModel, ArtifactModel
+from mlte.artifact.type import ArtifactType
+from mlte.evidence.metadata import EvidenceMetadata
+from mlte.value.model import OpaqueValueModel, ValueModel, ValueType
+
+
+def _has_callable(type, name) -> bool:
+    """Determine if `type` has a callable attribute with the given name."""
+    return hasattr(type, name) and callable(getattr(type, name))
+
+
+class Value(Artifact, metaclass=abc.ABCMeta):
```

Comment:
This looks mostly great, but I do have one question. I understand what you did, adding the abstract seriaize/deserialize methods for extensions to implement, and implementing to/from moodel by using Opaque and calling the abstract serialize/deserialize that an extended value should implement. All of that looks fine and makes sense. What I do not understand is why you created a new Value class, with the same name as the Value artifact, and copying/pasting all the code from the Value artifact class, instead of deriving from it. Again, I try to avoid having duplicated code when possible. Was it just so it could be called Value too? Why couldn't this be called something like BaseValue, and have it extend Value, and simply add the 2 abstract methods and the 2 implemented methods? Is it to avoid some of the polymorphism/model issues?

Reply:
The primary reason for this is the "contract" that the client of the type agrees to. We (developers of MLTE internals) override the `mlte.value.artifact` type; to do this, we have to provide `to_model()` and `from_model()` methods to satisfy the artifact protocol. On the other hand, external users of MLTE override the `mlte.value.base` type; to do this, they have to provide the `serialize()` and `deserialize()` methods. Moreover, this latter type already provides the `to_model()` and `from_model()` methods required by the artifact protocol. Ultimately, these two types exist at slightly different levels of abstraction.

I suppose that we could have this derive from the `Value` artifact, like you suggested, and call it `BaseValue`. The reason I didn't take this approach was because I don't want the user to ever have to know about the distinction between `artifact.value` and `base.value`. They are only ever exposed to `base.value` and don't have to worry about the extra work that we have to do to make the abstraction work.

Reply:
Ok, that is fine, I understand and agree with having the two Values. I also understand the reasoning for keeping its name Value, and that is fine too. But we could still derive Value from Value, and keep them with the same name, right? We would just need, inside base.value, to use full package names or alias to derive it from artifact.value. This would still be transparent to the user, but simplify maintainance effort by not duplicating the common code between base.value and artifact.value. This would still be hidden to the user.

Reply:
Agreed. I just tried this approach, and it does indeed work to have `base.Value` derive from `artifact.Value`. However, the only code duplication this removes are a couple of lines in the `__init__` method in `base.Value`. Other than that, there is little overlap between the two. The major implementation in `base.Value` are the `to_model()` and `from_model()` methods. These are not implemented by the `artifact.Value` class. Are there other bits of functionality that you are concerned about? Otherwise, I don't think its worth extending the inheritance hierarchy one level deeper to remove these three lines in the constructor (because it makes it slightly harder to reason about the types, in my opinion).

Reply:
Resolved in Slack.